### PR TITLE
CLI: Specify signing keys

### PR DESCRIPTION
**Pull request by MichelSantos**
PR for Issue #2048

### CODE REVIEW

**Comment by pmconrad** [comment]
Travis fails:
```
fatal error in "cli_sign_message": signal: SIGSEGV, si_code: 0 (memory access violation at address: 0x00000080)
/home/travis/build/bitshares/bitshares-core/tests/cli/main.cpp(1199): last checkpoint
```

**Comment by pmconrad** [comment]
I believe you are correct that the error is unrelated to your change.

File: tests/cli/main.cpp [hunk 1]
```diff
@@ -261,6 +254,7 @@ class client_connection
    fc::api<graphene::wallet::wallet_api> wallet_api;
    std::shared_ptr<fc::rpc::cli> wallet_cli;
    std::string wallet_filename;
+   fc::future<void> was_closed;
```
Comment (pmconrad) [inline tests/cli/main.cpp]: Please remove this, I believe it's a leftover from a previous fix attempt.

File: tests/cli/main.cpp [hunk 2]
```diff
@@ -495,6 +485,122 @@ BOOST_FIXTURE_TEST_CASE( cli_get_signed_transaction_signers, cli_fixture )
...
+      flat_set<public_key_type> actual_signers = con.wallet_api_ptr->get_transaction_signers(signed_trx);
+      BOOST_CHECK(signers == expected_signers);
+
+   } catch (fc::exception &e) {
+      edump((e.to_detail_string()));
+      throw;
+   }
+}
+
+
```
Comment (pmconrad) [inline tests/cli/main.cpp]: Perhaps add a test case that uses sign_builder_transaction2 to create a transaction authorized by a CAA?

File: tests/cli/main.cpp [hunk 3]
```diff
@@ -174,12 +174,51 @@ bool generate_block(std::shared_ptr<graphene::app::application> app, graphene::c
...
+//////
+uint32_t generate_blocks(std::shared_ptr<graphene::app::application> app, fc::time_point_sec timestamp)
+//uint32_t generate_blocks(fc::time_point_sec timestamp, bool miss_intermediate_blocks, uint32_t skip)
+{
+   bool miss_intermediate_blocks = true;
+   fc::ecc::private_key committee_key = fc::ecc::private_key::regenerate(fc::sha256::hash(string("nathan")));
+   uint32_t skip = ~0;
+   auto db = app->chain_database();
+
+   if( miss_intermediate_blocks )
```
Comment (pmconrad) [inline tests/cli/main.cpp]: Hm, this is always true? Is this intentional?

File: tests/cli/main.cpp [hunk 4]
```diff
@@ -1409,3 +1410,184 @@ BOOST_FIXTURE_TEST_CASE( general_storage, cli_fixture )
...
+
+      account_object nathan_acct = con.wallet_api_ptr->get_account("nathan");
+      INVOKE(upgrade_nathan_account);
+
+      // Register Alice account
+      const auto alice_bki = con.wallet_api_ptr->suggest_brain_key();
+      con.wallet_api_ptr->register_account(
+              "alice", alice_bki.pub_key, alice_bki.pub_key, "nathan", "nathan", 0, true
+      );
+      const account_object &alice_acc = con.wallet_api_ptr->get_account("alice");
```
Comment (pmconrad) [inline tests/cli/main.cpp]: `wallet::get_account` returns an account object, not a reference (unlike `database::get_account`).

File: tests/cli/main.cpp [hunk 5]
```diff
@@ -1409,3 +1410,184 @@ BOOST_FIXTURE_TEST_CASE( general_storage, cli_fixture )
...
+
+      // Restriction should have "to"  equal Charlie
+      vector<restriction> restrictions;
+      auto to_index = member_index<transfer_operation>("to");
+      restrictions.emplace_back(to_index, restriction::func_eq, charlie_acc.get_id());
+
+      con_alice.wallet_api_ptr->add_operation_to_builder_transaction(tx_alice_handle, caop);
+      asset ca_fee = con_alice.wallet_api_ptr->set_fees_on_builder_transaction(tx_alice_handle, GRAPHENE_SYMBOL);
+
+      // Sign the transaction with the implied nathan's key and the explicitly yet unnecessary Bob's key
```
Comment (pmconrad) [inline tests/cli/main.cpp]: The comment is wrong. A simple `sign_builder_transaction` should suffice here.

File: tests/cli/main.cpp [hunk 6]
```diff
@@ -1409,3 +1410,184 @@ BOOST_FIXTURE_TEST_CASE( general_storage, cli_fixture )
...
+      signed_trx = con_alice.wallet_api_ptr->sign_builder_transaction2(tx_alice_handle, {}, true);
+
+      // Check for one signatures on the transaction
+      BOOST_CHECK_EQUAL(signed_trx.signatures.size(), 1);
+
+      // Check that the signed transaction contains Bob's signature
```
Comment (pmconrad) [inline tests/cli/main.cpp]: Alice's

File: tests/cli/main.cpp [hunk 7]
```diff
@@ -1409,3 +1410,184 @@ BOOST_FIXTURE_TEST_CASE( general_storage, cli_fixture )
...
+      const asset transfer_amount = asset(123 * GRAPHENE_BLOCKCHAIN_PRECISION);
+      transfer_operation top;
+      top.from = alice_acc.id;
+      top.to = charlie_acc.id;
+      top.amount = transfer_amount;
+
+      con_bob.wallet_api_ptr->add_operation_to_builder_transaction(tx_bob_handle, top);
+      asset transfer_fee = con_bob.wallet_api_ptr->set_fees_on_builder_transaction(tx_bob_handle, GRAPHENE_SYMBOL);
+
+      // Sign the transaction with the implied nathan's key and the explicitly yet unnecessary Bob's key
```
Comment (pmconrad) [inline tests/cli/main.cpp]: I think you mean Alice's key is implied (but not available). Bob's key is required.

File: tests/cli/main.cpp [hunk 8]
```diff
@@ -1409,3 +1410,184 @@ BOOST_FIXTURE_TEST_CASE( general_storage, cli_fixture )
...
+      signed_trx = con_bob.wallet_api_ptr->sign_builder_transaction2(tx_bob_handle, {bob_bki.pub_key}, true);
+
+      // Check for two signatures on the transaction
```
Comment (pmconrad) [inline tests/cli/main.cpp]: ~~two~~one

File: tests/cli/main.cpp [hunk 9]
```diff
@@ -1409,3 +1410,184 @@ BOOST_FIXTURE_TEST_CASE( general_storage, cli_fixture )
...
+      BOOST_CHECK_EQUAL(bob_balances.size(), 0);
+
+      // Check Alice's balance
+      vector<asset> alice_balances = con.wallet_api_ptr->list_account_balances("alice");
+      BOOST_CHECK_EQUAL(alice_balances.size(), 1);
+      asset alice_core_balance = alice_balances.front();
+      asset expected_alice_balance =  asset(450000 * GRAPHENE_BLOCKCHAIN_PRECISION)
+                                      - expected_charlie_core_balance
+                                      - ca_fee - transfer_fee;
+      wdump((alice_core_balance)(expected_alice_balance)(expected_charlie_core_balance)(ca_fee)(transfer_fee));
```
Comment (pmconrad) [inline tests/cli/main.cpp]: Please remove this output.

**Comment by MichelSantos** [comment]
@pmconrad Thanks very much for your fix

**Pull request merged**